### PR TITLE
[PyCDE] Change module API to extend `Module` instead of using a decorator

### PR DESCRIPTION
**Pull request by teqdruid**
Refactor PyCDE module creation to the following syntax. This is a breaking API change, but a largely mechanical update to users code.

```py
class Mux(Module):
  clk = Clock()
  data = Input(dim(8, 14))
  sel = Input(types.i4)

  out = Output(types.i8)

  @generator
  def build(self):
    sel_reg = self.sel.reg()
    self.out = self.data.reg()[sel_reg].reg()
```

The previous approach had several issues:
- It wasn't possible for users to extend their modules by just subclassing them.
- Code completion for ports inside of generators didn't work.
- We are going to introduce functional models, wherein a method of a model will be interacting with the instance. We wanted hardware generation to have a similar feeling.
- Internally, supporting more module like constructs (e.g. ESI service implementations) involved specifying a bunch of callbacks. This replaces that with an OOP approach.
- The new code is far better structured (and documented).

Much as I tried to not boil the ocean, I ended up boiling a small pond. AFAICT, this cannot be split up into multiple PRs. It also appears larger than it is since I had to update all the tests and git got confused and thinks there's a bunch more new code than there actually is.

### CODE REVIEW

**Comment by teqdruid** [comment]
For the lurkers and future reviewers: this is one of the refactoring changes I've been meaning to make for some time now. We've had this problem of multiple things which aren't strictly modules wanting to be able to generate like modules with a few extras. Previously, we were accomplishing this with ad-hoc callbacks. This change makes it possible to customize module types by just subclassing `Module`, which is something we'll be doing more in the future.

It does break existing code and touched a _lot_. I'm still fiddling with it, cleaning it up, and have yet to document it. However, I think it is a significant improvement in code organization. Feels less ad-hoc now. Also, for (essentially) the same functionality, it is a LoC reduction!

**Comment by teqdruid** [comment]
I tried many alternatives to metaclasses since:

“Metaclasses are deeper magic than 99% of users should ever worry about. If you wonder whether you need them, you don’t (the people who actually need them know with certainty that they need them, and don’t need an explanation about why).”

— Tim Peters

but in this case, I found them to accomplish exactly what I wanted.

**Comment by teqdruid** [comment]
Even with a massive amount more documentation added and git thinking a bunch of old code is new, this refactoring is still a net negative LoC diff!

File: frontends/PyCDE/src/module.py [hunk 1]
```diff
@@ -53,279 +53,450 @@ def val_str(val):
...
 
-def generate_msft_module_op(generator: Generator, spec_mod: _SpecializedModule,
-                            **kwargs):
 
-  def create_output_op(args: _GeneratorPortAccess):
-    """Create the hw.OutputOp from module I/O ports in 'args'."""
-    output_ports = spec_mod.output_ports
-    outputs: list[Value] = list()
+class _BlockContext:
+  """Bookkeeping for a scope."""
```
Comment (mortbopet) [inline frontends/PyCDE/src/module.py:70]: nit: A scope in what/where?

File: frontends/PyCDE/src/module.py [hunk 2]
```diff
@@ -53,279 +53,450 @@ def val_str(val):
...
+    _current_block_context.reset(self._old_system_token)
+
+  def uniquify_symbol(self, sym: str) -> str:
+    """Create a unique symbol and add it to the cache. If it is to be preserved,
+    the caller must use it as the symbol on a top-level op."""
+    ctr = 0
+    ret = sym
+    while ret in self.symbols:
+      ctr += 1
+      ret = sym + "_" + str(ctr)
```
Comment (mortbopet) [inline frontends/PyCDE/src/module.py:98]: meganit: `ret = f"{sym}_{ctr}"`

File: frontends/PyCDE/src/module.py [hunk 3]
```diff
@@ -53,279 +53,450 @@ def val_str(val):
...
+    damage."""
+
+    self._block_args = None
+    self._output_values = None
+    self._builder = None
+
+
+class ModuleLikeBuilderBase(_PyProxy):
+  """`ModuleLikeBuilder`s are responsible for preparing `Module` and ilk
+  subclasses for use. They are responsible for scanning the subclass' attribute,
```
Comment (mortbopet) [inline frontends/PyCDE/src/module.py:173]: what are `ilk subclasses`?

**Review by mortbopet — APPROVED** [review]
The callbacks always seemed like a hack to me, so is a nice evolution of the project - as well as a big +1 for all of the added documentation. A few nits, but seeing as the PyCDE tests pass I'm going to blindly trust that this refactor didn't introduce any regressions.

**Comment by teqdruid** [comment]
> The callbacks always seemed like a hack to me, so is a nice evolution of the project - as well as a big +1 for all of the added documentation. A few nits, but seeing as the PyCDE tests pass I'm going to blindly trust that this refactor didn't introduce any regressions.

I forgot to mention: I broke the FSM construct support. There is a straightforward path to fixing it, but I didn't want to add more code to this PR, so I just XFAIL'd the test.

**Comment by mikeurbach** [comment]
This is awesome, big +1 from me as well. We did some neat things with the decorators, but using Python inheritance seems much more natural in hindsight. I'm glad you were able to evolve things in this direction and solve the problems you listed in the PR description. I haven't reviewed the implementation, but I'll take a look now, since I'm genuinely curious how you pulled this off.

**Pull request merged**